### ISSUE TITLE
QUERY: `stats.wasserstein_distance`: adding term changes pytorch model backpropagation at large scales

### ISSUE THREAD

**Issue by AymenTlili131** [scipy/scipy#21978]
Hi I'm training a model in pytorch using the scipy wasserstein distance added to the pytorch mean squared error loss which has a computation graph attached to it dicating which neurones get updated with what amount of gradients .

Adding the scipy term should not create a different gradient since it is not derivable by pytorch and can't stand the Loss.backward() on it's own but it does if added to MSE.

My issue here is figuring out why it's changing the gradient after a certain number of steps leading to a completly different model . The issue has been up on pytorch forums as well and I hope for an explination . Here's the link for extended version with images https://discuss.pytorch.org/t/adding-non-derivable-wasserstein-from-scipy-to-pytorch-mse-works/213191/2

```
import torch
from torch.utils.data import Dataset, DataLoader
import torch.nn as nn
import torch.optim as optim
import numpy as np
import random
import copy

seed=74
torch.manual_seed(seed)
np.random.seed(seed)
random.seed(seed)


# Create a PyTorch-ready dataset
class ToyRegressionDataset(Dataset):
    def __init__(self, n_samples=10000, noise=0.1):
        # Generate random input data (features)
        self.x = torch.linspace(-10, 10, n_samples).unsqueeze(1)  # Shape: (n_samples, 1)
        # Generate labels (targets) based on the quadratic equation with noise
        self.y = 3 * self.x**2 + 2 * self.x + 1 + noise * torch.randn_like(self.x)
    
    def __len__(self):
        return len(self.x)
    
    def __getitem__(self, index):
        return self.x[index], self.y[index]

# Instantiate the dataset and dataloader
dataset = ToyRegressionDataset()
dataloader = DataLoader(dataset, batch_size=25, shuffle=True)

# Define a simple regression model
class RegressionModel(nn.Module):
    def __init__(self):
        super(RegressionModel, self).__init__()
        self.model = nn.Sequential(
            nn.Linear(1, 64),
            nn.ReLU(),
            nn.Linear(64, 32),
            nn.ReLU(),
            nn.Linear(32, 1)
        )
    
    def forward(self, x):
        return self.model(x)
```
and the cell that mixes it all (MSE + regularizing ws to scame scale as MSE X ws)

```
# Initialize the model, loss function, and optimizer
model_ws_reg=copy.deepcopy(model1)
criterion = nn.MSELoss()
optimizer = optim.Adam(model_ws_reg.parameters(), lr=0.01)

# Training loop
n_epochs = 1001
for epoch in range(n_epochs):
    b_i=0
    for x_batch, y_batch in dataloader:
        # Forward pass
        predictions = model_ws_reg(x_batch)
        loss = criterion(predictions, y_batch)
        vec1_np = predictions.detach().cpu().numpy()
        vec2_np = y_batch.detach().cpu().numpy()

        wsd_list = [wasserstein_distance(vec1_np[i], vec2_np[i]) for i in range(vec1_np.shape[0])]
        # If you need an aggregate measure, you can compute the average distance
        average_wsd = sum(wsd_list) / len(wsd_list)
        loss = criterion(predictions,y_batch)+average_wsd*0.0001
        # Backward pass and optimization
        if b_i==0 and epoch==0 :
          for name, param in model_ws_reg.named_parameters():
            if param.grad is not None:
                print(f"Parameter: {name} {param}, Gradient: {param.grad[:10]}")
                break
        optimizer.zero_grad()
        loss.backward()
        optimizer.step()
        if b_i==0 and epoch==0 :
          for name, param in model_ws_reg.named_parameters():
            if param.grad is not None:
                print(f"Parameter: {name} {param}, Gradient: {param.grad[:10]}")
                break
        b_i=b_i+1
    if (epoch) % 50 == 0:
        print(f"Epoch [{epoch+1}/{n_epochs}], Loss: {loss.item():.4f}")
for name, param in model.named_parameters():
  if param.grad is not None:
    print(f"Parameter: {name} {param}, Gradient: {param.grad[:10]}")
    break
# Test the model on new data
x_test = torch.linspace(-10, 10, 10000).unsqueeze(1)
y_test = 3 * x_test**2 + 2 * x_test + 1
y_pred = model_ws_reg(x_test).detach()

# Plot the results
import matplotlib.pyplot as plt

plt.figure(figsize=(8, 6))
plt.scatter(dataset.x, dataset.y, label="Original Data", color="blue", alpha=0.5)
plt.plot(x_test, y_test, label="True Function", color="green")
plt.plot(x_test, y_pred, label="Model Prediction", color="red")
plt.legend()
plt.title("Regression Model Fit")
plt.show()
The print outputs are as follows (from left to right : MSE , MSE+ws_scipy , MSE+0.0001ws_scipy )
for 1000 data points and a batch size of 2 intended to create many training steps the results are identical and 100 epochs .
```

in the example i’m running now i’ve tried to increase those by 10s or 100 folds but even with that and for context key dissimalirities between this and working on a smaller batch size and x10 less epochs . 

the toy example is trying to fit this curve but I don’t know if the scale of values is comparable but i understand your motivation to showcase the change happening after one step , which there is none of here

Hope the info helps and gives more context

update after the increases 100folds : the code I’m posting here is about 1000 epochs instead and here is the same comparison left to right . I’m assuming tracking the first named neurone is enough as params are put in an ordered dict data structure to my knowledge,correct me if not.
(note also i didn’t enforce my package versions to deny that this problem stems from a certain old interaction that might’ve already been debugged)

so the end model is the same but u can notice at the start of training that the gradients were different . I think the fact that the toy example is simple led to this unique solution

**Comment by AymenTlili131** [comment]
Hi 
adding `model1= RegressionModel() ` makes the code executable just before a deep copy of it . please check the link i provided .

`loss = criterion(predictions, y_batch)  ` works normally and that's pytorch normal training on Mean Squared Error criterion

`loss = criterion(predictions,y_batch)+average_wsd*0.0001` is adding a pytorch criterion term that has .backward() method implemented and the scypy term which doesn't . surprisingly the resulting tensor works and does support .backward() .

One would assume that the scipy just added to the norm of the single valued tensor and not touch the gradients and backward() would only move the pytorch related information .

if you try this experiment with batch_size=2 and epochs=100 vs batch size=25 and epochs=1000 .you will notice as stated before that only during the second experient do _gradients change_ WHEN then **shouldn't** .

the difference is not negligeable even in this toy example and in my own large scale experiment the addition of this term gave better results . refer to the link i provided to see gradients' first components varying from -14 ,-28 ,-15 in 3 experiments that all started with deep copy differences being the Loss used . MSE , MSE+ws_scipy , MSE+ws_scipy*0.0001 

Thanks for the quick answer

**Comment by mdhaber** [comment]
@AymenTlili131 IIUC, I think the answer to my question:

> What happens if you change it to: `loss = criterion(predictions, y_batch)`

is:

> `loss = criterion(predictions, y_batch)` works normally and that's pytorch normal training on Mean Squared Error criterion

If that's the case, then think this issue can be closed. `average_wsd` is just a NumPy array; the fact that it happened to be calculated by SciPy is not what is changing the results of your calculation. Although `wasserstein_distance` will soon support alternative backends, `wasserstein_distance` currently accepts only NumPy arrays as input (or converts inputs to NumPy arrays) and produces NumPy arrays as output. It does not advertise the capability to work seamlessly with Tensors and backpropagation, so while that may be a feature request for this and many SciPy functions (tracked elsewhere), the lack of support for Tensors and backpropagation is not a bug. The other possibility for a bug was if just calling `wasserstein_distance` with your input - without even including `average_wsd` in the `loss` - was somehow changing your results (e.g. by modifying an input array), but that doesn't appear to be happening.

This forum is for reporting bugs and requesting enhancements; for usage questions, please consider Stack Overflow (or these days, generative AI). If I've missed something, please point out how how SciPy's itself is the source of the trouble and we can reopen.

**Comment by AymenTlili131** [comment]
Hi sorry but I still don't think you grasp the issue at hand .
```
        predictions = model_ws_reg(x_batch)
        vec1_np = predictions.detach().cpu().numpy()
        vec2_np = y_batch.detach().cpu().numpy()
        wsd_list = [wasserstein_distance(vec1_np[i], vec2_np[i]) for i in range(vec1_np.shape[0])]
        average_wsd = sum(wsd_list) / len(wsd_list)
        loss = criterion(predictions,y_batch)+average_wsd
```
in lines 2 and 3 in this snippet of my code i purposefully make my inputs numpy arrays to match your API and i fully knowingly add the output numpy array of the result  to a pytorch tensor in the last line 

> wasserstein_distance currently accepts only NumPy arrays as input (or converts inputs to NumPy arrays) and produces NumPy arrays as output.

`loss = criterion(predictions, y_batch)` was only commented in to **show you the benchmark experience** that made the difference  and is the expected outcome to be identical with if your claim holds true.The existence of a difference disproved the claim that your implementation does not affect backpropagation (even with no attached backpropagation graph or even the same data structure type) 

> It does not advertise the capability to work seamlessly with Tensors and backpropagation

average_wsd SHOULD NOT AFFECT THE TRAINING because **you implemented it not to** and it lacks backpropagation capabilities  . this entire post is highlighting **an Experiment where IT DOES** contribute to training by creating different gradients than the loss that doesn't include the term .THIS IS AN _UNINTENDED FEATURE and a BUG_ which i'm seeking your help to make sense of  . 


When actually using   `loss = criterion(predictions,y_batch)+average_wsd`  , you would find that your implementation is affecting things it shouldn't compared to the use of `loss = criterion(predictions,y_batch)`

as the the second time urging you to look at the link I provided,I gladly bring the key screenshot here for your ease of access .

**EVIDENCE** (LOOK AT THE FIRST LINE/GRADIENTS OF EACH EXPERIMENT, THE GRADIENTS ARE DIFFERENT ! ) 
![image](https://github.com/user-attachments/assets/3a5dc359-0e68-42b6-94b4-74784227db3c)
Figure : comparing the same value of a neurone AND **IT'S GRADIENT** changing when introducing wasserstein distance that does not support backpropogation  (WHEN it's intended not to change them)

AND THIS IS me printing the gradients of the **first neurone only** of identical input data , recreation seeds , pytorch initialized model  and the same hyper parameters . A DENSE LAYER has many more of these neurones than what I prented and the fact your term is changing things so much in even one neurone,should give you an idea of what happens in an entire Layer and how you can erroneously mess entire models but I managed to bring you an example which within not even one epoch of learning gives you a different result than what you would expect it to so IT IS A **BUG**

(my original issue is on a much larger architecture but I managed to narrow things down to this term by using this toy example) 

The fact that the loss term value changes between experiments is normal , but only the pytorch part of that loss should be propagated and only should be producing gradients hence the model after a single step or a thousand should always be the same . but the different gradients mean that given complex enough inputs , the each instance of these 3 models would converge to something entirely different from the other . the toy example happened to be simple enough that despite the different gradients we arrive at the same neuron value with same gradients as if there is a unique solution  . the github i posted in that link would show you that with complex enough data as input the changing gradient does have noticeable effects 

I'm giving you another piece of evidence training a much bigger model with your loss and the geomloss implementation equivalent intended to work with pytorch and for something that the presence of should not contribute anything , it has achieved an average +20% accuracy in on other methods .
 
![image](https://github.com/user-attachments/assets/ae4ccd51-f659-4228-94ec-ae9f1f080f18)

please, i'm not a beginner using these tools but in face of this challenging bug , please treat me like one and be patient looking at the evidence I provide and try to run the code changing the different varients of the code to replicate the screenshots. I gave you the regularized version of the toy example so that I don't post 3 identical blocks of code where the only difference is removing the regularizing 0.0001 in one experience and removing the entire wassertstein_distance  term in the second

At a large enough scale , launching multiple instances of the same model to train with identical losses (MSE) then adding `wasserstein_distance` or a regularized amount of it (COMPUTED FOR YOUR COMPATIBILITY ON numpy array versions of batches ) should , if your claim holds true , yield the same results as your distance does NOT contribute to training so WHY is it's presence is causing pytorch to create different gradients  ? 
I know you didn't implement pytorch's autograd but i'm just asking from your perspective why it's interacting with it this way.

Thanks and hope the lengthy explanation contributes to upholding scipy's standards

**Comment by ev-br** [comment]
Please avoid yelling at people with all_caps.

**Comment by mdhaber** [comment]
Please help us help you by reducing your notebook to one relevant cell that is as short as possible. 

- Running the cell should produce minimal output capable of showing the problem (e.g. the first few components of the gradient), then stop.
- Running the cell multiple times without changing the code should produce identical output.
- Changing the line `loss = criterion(predictions,y_batch) + average_wsd` should demonstrate that the addition of `average_wsd`, which should be a constant, changes the gradient. Adding a fraction of `average_wsd` should change the gradient less, if I understand correctly.

Please also make your notebook view-only, allowing others to execute the code and make temporary changes, but changes being saved to your notebook.

**Comment by WarrenWeckesser** [comment]
Quick comment and debugging suggestion: `average_wsd` will be a numpy scalar, specifically `np.float64`, so, as noted above, this issue appears to have nothing to do with SciPy, and it is appropriate that the issue was closed.  If adding a numpy scalar to a PyTorch tensor causes a problem with the gradient computation, then it is likely an issue in PyTorch.

@AymenTlili131, you might be able to verify this by changing the computation of `loss` to

        loss = criterion(predictions,y_batch) + np.float64(0.5)

and maybe also

        loss = criterion(predictions,y_batch) + 0.5

to see if either of those results in the same issue that you have observed.

**Comment by AymenTlili131** [comment]
Hi everyone 
I'd like to start off by apologizing for my tone which @ev-br pointed out . it's uncouth to be using caps while asking for help .
I posted here as per my research advisor instruction even tho I fully know that your implementation has nothing to do with the aberrant results. That being said the early closing of the issue came as a cause for not providing the necessary materials and explain my problem well .
my advisor was observing this issue and accusing me of falsifying research results if i didn't prove that wasserstein_distance was affecting training and I didn't contact you about it and not making the issue clear to you would cost me my job even knowing that your code is working as intended hence coercing me to answer emotionally because my livelihood is on the line. even tho I reran the experiment several times and with wasserstein giving an advantageous identical result each time thanks to version control tools I setup in place to replicate it.
mentionned here : https://discuss.pytorch.org/t/adding-non-derivable-wasserstein-from-scipy-to-pytorch-mse-works/213191 
maintained here : https://github.com/AymenTlili131/Federated-Continual-learning-

As per @mdhaber's request I took the collab  and executed a singuler cell changing the loss and resetting the runtime each time . the results were identical hence your implementation is working as expected . running 3 cells back to back each with a different loss however gave something differing each time .

This indicates that something wasn't being reset despite me redeclaring a model from a deepcopy, an optimizer and traversing the same data in the same order.but this is outside the of the issue and i apologize again for the inconvenience.

@WarrenWeckesser That is not the exact setting of the issue as the second term changes and just replacing it with a constant might not have the same effect

Sorry to have wasted your time and thank you for your patience . rerunning my original github bigger experiment still yields different results each time ,I just don't know why that it is yet .No one can help me except myself so i'll just have to be better and figure it out like I always did on my own .

You may close this issue
